File: python/merge_strings_alternately.py
class Solution:
    def mergeAlternately(self, word1: str, word2: str) -> str:
        w1: int = 0
        w2: int = 0
        result: str = ""
        while w1 < len(word1) and w2 < len(word2):
            result += word1[w1]
            result += word2[w2]
            w1 += 1
            w2 += 1

        while w1 < len(word1):
            result += word1[w1]
            w1 += 1

        while w2 < len(word2):
            result += word2[w2]
            w2 += 1

        return result


def test_mergealternately_case_1():
    # arrange
    word1: str = "abc"
    word2: str = "pqr"
    expected: str = "apbqcr"

    # act
    solution = Solution()
    actual = solution.mergeAlternately(word1, word2)

    # assert
    assert expected == actual


def test_mergealternately_case_2():
    # arrange
    word1: str = "ab"
    word2: str = "pqrs"
    expected: str = "apbqrs"

    # act
    solution = Solution()
    actual = solution.mergeAlternately(word1, word2)

    # assert
    assert expected == actual


def test_mergealternately_case_3():
    # arrange
    word1: str = "abcd"
    word2: str = "pq"
    expected: str = "apbqcd"

    # act
    solution = Solution()
    actual = solution.mergeAlternately(word1, word2)

    # assert
    assert expected == actual

File: rust/src/code/merge_strings_alternately.rs
struct Solution {}

impl Solution {
    pub fn merge_alternately(word1: String, word2: String) -> String {
        let mut w1: usize = 0;
        let mut w2: usize = 0;
        let word1: Vec<char> = word1.chars().collect();
        let word2: Vec<char> = word2.chars().collect();
        let mut result: String = "".to_string();

        while w1 < word1.len() && w2 < word2.len() {
            result.push(word1[w1]);
            result.push(word2[w2]);
            w1 += 1;
            w2 += 1;
        }

        while w1 < word1.len() {
            result.push(word1[w1]);
            w1 += 1;
        }

        while w2 < word2.len() {
            result.push(word2[w2]);
            w2 += 1;
        }

        result
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_merge_alternately_case_1() {
        // arrange
        let word1: String = "abc".to_string();
        let word2: String = "pqr".to_string();
        let expected: String = "apbqcr".to_string();

        // act
        let actual = Solution::merge_alternately(word1, word2);

        // assert
        assert_eq!(expected, actual);
    }
}
